waddrmgr/scoped_manager.go

```
@@ -2014,6 +2014,74 @@ func (s *ScopedKeyManager) ImportPublicKey(ns walletdb.ReadWriteBucket,
+// account state and extended public keys, avoiding database access.
+func (s *ScopedKeyManager) DeriveAddr(account uint32, branch uint32,
+	index uint32) (btcutil.Address, []byte, error) {
+
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	acctInfo, ok := s.acctInfo[account]
+	if !ok {
+		return nil, nil, managerError(ErrAccountNotCached,
+			"account not cached", nil)
+	}
+
+	return s.deriveAddr(acctInfo, account, branch, index)
+}
+
+// DeriveAddrs derives a range of addresses and their corresponding pkScripts
+// for the given account and branch. It generates `count` addresses starting
+// from `startIndex`. This method relies on the in-memory account state and
+// extended public keys, avoiding database access.
+//
+// It returns:
+// - A slice of derived addresses.
+// - A slice of corresponding pkScripts.
+// - An error if the account is not cached or derivation fails.
+func (s *ScopedKeyManager) DeriveAddrs(account uint32, branch uint32,
+	startIndex uint32, count uint32) ([]btcutil.Address, [][]byte, error) {
+
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	// Ensure the account information is cached. If not, we cannot proceed
+	// without a DB transaction, so we return an error. The caller is
+	// expected to ensure the account is loaded (e.g. via AccountProperties)
+	// before calling this method.
+	acctInfo, ok := s.acctInfo[account]
+	if !ok {
+		return nil, nil, managerError(ErrAccountNotCached,
+			"account not cached", nil)
+	}
+
+	addrs := make([]btcutil.Address, 0, count)
+	scripts := make([][]byte, 0, count)
+
+	// Iterate through the requested range of child indexes (startIndex to
+	// startIndex+count). For each index, we derive the corresponding
+	// extended key and convert it into a payment address and script.
+	//
+	// TODO(yy): Optimize by deriving the branch key once outside the loop
+	// instead of re-deriving it for every index via s.deriveKey.
+	endIndex := startIndex + count
```

Should we add a guard to prevent `uint32` overflow? For example, if the current index is close to the maximum value, say max minus 10, and we try to derive 20 entries, we would end up with an invalid derivation. Adding an explicit check could help fail early and avoid subtle bugs.

---

nice catch - added!